src/o365/spo/commands/tenant/tenant-recyclebinitem-restore.ts

```diff
@@ -0,0 +1,203 @@
+import commands from '../../commands';
+import SpoCommand from '../../../base/SpoCommand';
+import GlobalOptions from '../../../../GlobalOptions';
+import { CommandOption, CommandCancel, CommandValidate } from '../../../../Command';
+
+const vorpal: Vorpal = require('../../../../vorpal-init');
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  url: string;
+  wait?: boolean;
+}
+
+interface RestSpoOperation {
+  HasTimedout: boolean;
+  IsComplete: boolean;
+  PollingInterval: number;
+}
+
+class SpoTenantRecycleBinItemRestoreCommand extends SpoCommand {
+  private timeout?: NodeJS.Timer;
+  private readonly maxAttempts: number = 5;
+
+  public get name(): string {
+    return commands.TENANT_RECYCLEBINITEM_RESTORE;
+  }
+
+  public get description(): string {
+    return 'Restores the specified deleted site collection from tenant recycle bin';
+  }
+
+  public getTelemetryProperties(args: CommandArgs): any {
+    const telemetryProps: any = super.getTelemetryProperties(args);
+    telemetryProps.url = typeof args.options.url !== 'undefined';
+    telemetryProps.wait = typeof args.options.wait !== 'undefined';
+    return telemetryProps;
+  }
+
+  public commandAction(cmd: CommandInstance, args: CommandArgs, cb: (err?: any) => void): void {
+    const wait: boolean = args.options.wait || false;
+    let spoAdminUrl: string;
+
+    this.getSpoAdminUrl(cmd, this.debug)
+    .then((adminUrl: string): Promise<RestSpoOperation> => {
+      spoAdminUrl = adminUrl;
+      return this.restoreDeletedSite(args.options.url, spoAdminUrl, cmd);
+    })
+    .then((response: RestSpoOperation): Promise<void> => {
+      if (!response.HasTimedout && response.IsComplete) {
+        if (this.verbose) {
+          cmd.log('site collection restored');
+        }
+
+        return Promise.resolve();
+      }
+      else if (wait) {
+        return new Promise<void>((resolve: () => void, reject: (error: any) => void): void => {
+          this.waitForRestoreDeletedSite(
+            cmd,
+            this,
+            spoAdminUrl,
+            args.options.url,
+            resolve,
+            reject,
+            0
+          );
+        });
+      }
+      else {
+        return Promise.reject('site collection has not been restored');
+      }
+    })
+    .then(() => {
+      if (this.verbose) {
+        cmd.log(vorpal.chalk.green('DONE'));
+      }
+
+      cb()
+    }, (err: any): void => this.handleRejectedODataJsonPromise(err, cmd, cb));
+  }
+
+  public cancel(): CommandCancel {
+    return (): void => {
+      if (this.timeout) {
+        clearTimeout(this.timeout);
+      }
+    }
+  }
+
+  private waitForRestoreDeletedSite(cmd: CommandInstance, command: SpoTenantRecycleBinItemRestoreCommand, spoAdminUrl: string, siteToRestoreUrl: string, resolve: () => void, reject: (error: any) => void, iteration: number): void {
+    iteration++;
+
+    new Promise((): Promise<void> => {
```

Comment:
There is no need to wrap this code block in a Promise since it's neither returned nor used.